File: Algorithms/1695_Maximum_Erasure_Value/Python/Maximum_Erasure_Value_Solution_1.py


# Space: O(n)
# Time: O(n)

class Solution:
    def maximumUniqueSubarray(self, nums):
        length = len(nums)
        if length == 1: return nums[0]
        fast, slow, running_total, res, cache = 0, 0, 0, 0, set()

        while fast < length:
            if nums[fast] in cache:
                res = max(res, running_total)
                running_total -= nums[slow]
                cache.remove(nums[slow])
                slow += 1
            else:
                running_total += nums[fast]
                cache.add(nums[fast])
                fast += 1

        res = max(res, running_total)
        return res





        

File: Algorithms/1695_Maximum_Erasure_Value/Java/MaximumErasureValueSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int maximumUniqueSubarray(int[] nums) {
        if (nums.length == 1) return nums[0];
        int slow = 0, fast = 0, runningTotal = 0, res = 0;
        Set<Integer> cache = new HashSet<>();
        while (fast < nums.length) {
            if (cache.contains(nums[fast])) {
                res = Math.max(res, runningTotal);
                cache.remove(nums[slow]);
                runningTotal -= nums[slow];
                slow += 1;
            } else if (!cache.contains(nums[fast])) {
                cache.add(nums[fast]);
                runningTotal += nums[fast];
                fast += 1;
            }
        }
        res = Math.max(res, runningTotal);
        return res;
    }
}





